### PR TITLE
Remove duplicate default target (redux)

### PR DESCRIPTION
**Pull request by jyn514**
Same as #534 but now it doesn't crash when viewing source code.

r? @pietroalbini

### CODE REVIEW

**Comment by pietroalbini** [comment]
There are merge conflicts.

**Comment by pietroalbini** [comment]
...and it'd be nice to add tests, at least for the web part.

**Comment by jyn514** [comment]
> There are merge conflicts.
Yeah ... I tried to rebase, but it didn't work very well, I think because of the commit to master reverting the earlier changes. I'll fix the conflicts (again) once I've written some tests, probably tomorrow.

**Comment by jyn514** [comment]
So ... I can add tests for the web changes once #553 is merged (it has the `.file()` addition to FakeReleases). However, we don't currently have a way to test the build step works properly and generates all the static files it should. I might spend some time adding a FakeBuild tomorrow, but my schedule is about to get busy again so I may not finish.

**Comment by pietroalbini** [comment]
> However, we don't currently have a way to test the build step works properly and generates all the static files it should. I might spend some time adding a FakeBuild tomorrow, but my schedule is about to get busy again so I may not finish.

I'd not handle builds along with other tests in the codebase, but I'd do something separate (like rustwide's buildtest or crater's minicrater). My worry with adding built tests along with the other tests is that it encourages using a "fake build" around the codebase tests: builds are inheritely slow, as the rustwide initialization can't be parallelized, so we want to have as few of them as possible.

**Comment by jyn514** [comment]
This is ready for review.

File: src/db/migrate.rs [hunk 1]
```diff
@@ -269,6 +269,19 @@ fn migrate_inner(version: Option<Version>, conn: &Connection, apply_mode: ApplyM
             // downgrade query
             "DROP TABLE blacklisted_crates;"
         ),
+        migration!(
+            context,
+            // version
+            7,
```
Comment (pietroalbini) [inline src/db/migrate.rs]:
```suggestion
            8,
```

Migration 7 is already taken by https://github.com/rust-lang/docs.rs/pull/567.

File: src/web/rustdoc.rs [hunk 2]
```diff
@@ -224,6 +225,19 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
     req_path.insert(1, &name);
     req_path.insert(2, &version);
 
+    // if visiting the full path to the default target, remove the target from the path
+    // expects a req_path that looks like `/rustdoc/:crate/:version[/:target]/.*`
+    let crate_details = cexpect!(CrateDetails::new(&conn, &name, &version));
+    if req_path[3] == crate_details.metadata.default_target {
```
Comment (pietroalbini) [inline src/web/rustdoc.rs]: Can we add a `&& req_path.len() > 4`? Just to be safe, I don't want this to panic.

Comment (jyn514) [reply src/web/rustdoc.rs]: It works out, you can slice one past the bounds of the array without a problem: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ff716f47089c50c73b45839444f2a39b

File: src/db/migrate.rs (hunk 1 above)
Comment (jyn514) [reply src/db/migrate.rs]: This required a rebase.

File: src/web/rustdoc.rs (hunk 2 above)
Comment (jyn514) [reply src/web/rustdoc.rs]: I added a test just to be on the safe side.

**Comment by jyn514** [comment]
Addressed review comments.

**Pull request merged**